### ISSUE TITLE
RFC: `graphql.frag` helper

### ISSUE THREAD

**Issue by benjie** [0no-co/gql.tada#123]
## Summary

Status quo:

```ts
const FooFrag = graphql(`
  fragment FooFrag on Foo {
    __typename
  }
`);
function fn(data: FragmentOf<typeof FooFrag>) {
  const foo = readFragment(FooFrag, data);
  // ...
}
const Bar = graphql(`
  fragment BarFrag on Bar {
    foo {
      ...FooFrag
    }
  }
`, [FooFrag])
```

Remembering `FragmentOf<typeof ...>`, `readFragment(X, Y)`, etc is not immediately intuitive. It would be nice to have a fragment-centric API, where you start by writing the fragment variable (`FooFrag`) and use auto-complete from there:

```ts
const FooFrag = frag(`
  fragment FooFrag on Foo {
    __typename
  }
`);
function fn(data: typeof FooFrag.Data) {
  const foo = FooFrag.read(data);
  // ...
}
const Bar = frag(`
  fragment BarFrag on Bar {
    foo {
      ...FooFrag
    }
  }
`, [FooFrag])
```

## Proposed Solution

A `graphql.frag` helper function, something along the lines of:

```ts
interface GraphQLTadaAPI<Schema extends IntrospectionLikeType, Config extends AbstractConfig> {
  frag<
    const In extends stringLiteral<In>,
    const Fragments extends readonly [...makeDefinitionDecoration[]],
  >(
    input: In,
    fragments?: Fragments
  ): {
    Document: 
      getDocumentNode<
        parseDocument<In>,
        Schema,
        getFragmentsOfDocumentsRec<Fragments>,
        Config['isMaskingDisabled']
      >
    Data: FragmentOf<
      getDocumentNode<
        parseDocument<In>,
        Schema,
        getFragmentsOfDocumentsRec<Fragments>,
        Config['isMaskingDisabled']
      >
    >;
    read<const Fragment extends fragmentOfTypeRec<
      getDocumentNode<
        parseDocument<In>,
        Schema,
        getFragmentsOfDocumentsRec<Fragments>,
        Config['isMaskingDisabled']
      >
    >>(
      fragment: Fragment
    ): fragmentOfTypeRec<
      getDocumentNode<
        parseDocument<In>,
        Schema,
        getFragmentsOfDocumentsRec<Fragments>,
        Config['isMaskingDisabled']
      >
    > extends Fragment
      ? never
      : mirrorFragmentTypeRec<
          Fragment,
          ResultOf<
            getDocumentNode<
              parseDocument<In>,
              Schema,
              getFragmentsOfDocumentsRec<Fragments>,
              Config['isMaskingDisabled']
            >
          >
        >;
  };

```

(More work would be needed to allow `graphql(X, [...fragments])` to accept these wrapper objects in addition to the fragment documents they already accept.)

**Comment by kitten** [comment]
There's a couple of problems I can foresee here, that may stop this from being a viable option 🤔

Currently, there's a small conflict here with persisted documents (#77). The principle of persisted documents is that we'd only refer to documents by type and hence they'd treeshake/minify away entirely from the output bundle, and at that point only the “stand-in” root document would be left.
This still requires a change in `readFragment()` to `readFragment<typeof document>(data)`, but otherwise this is good to go.

The above proposal however is potentially incompatible with this idea and would reqire us to make stand-ins for fragments or define fragments in kind of awkward ways.

The other problem I'd have with this is that `typeof FooFrag.Data` would require `Data` (or whichever property name we'd choose for it) to be a required property to avoid `undefined` to be part of its type (otherwise type helpers would again be necessary).
This means that TypeScript would suddenly claim that `FooFrag.Data` is an actual property that exists on the wrapper object.

This could of course be mitigated by placing a getter on it that throws an error at runtime, but I still feel a little iffy about that.

In short, though, I'm not 100% convinced that:
- `Fragment.read(data)` over `readFragment(Fragment, data)`,
- and; `typeof Fragment.data` over `typeof FragmentOf<Fragment>`
are big enough wins in terms of usability/DX to justify this 🤔 Especially since it will increase the API surface area and double the APIs to facilitate this and introduce an additional wrapper type for fragments.

**Comment by benjie** [comment]
Fair comments. I'm still not happy with the existence of `typeof` even in my proposal; really I want them to be like classes so you can refer to the variable as both a variable (document) and a type (data type). I feel like there should be an ergonomic win here somewhere, but clearly it's not quite this proposal... I shall keep pondering on it :+1:
